asmcomp/riscv/emit.mlp

```diff
@@ -0,0 +1,693 @@
+        | {loc = Stack s; typ = (Val | Int | Addr)}, {loc = Reg _} ->
+            let ofs = slot_offset s (register_class src) in
+            emit_load dst ofs
+        | {loc = Stack s; typ = Float}, {loc = Reg _} ->
+            let ofs = slot_offset s (register_class src) in
+            emit_float_load dst ofs
+        | _ ->
+            Misc.fatal_error "Emit: Imove"
+      end
+  | Lop(Iconst_int n) ->
+      `	li	{emit_reg i.res.(0)}, {emit_nativeint n}\n`
+  | Lop(Iconst_float f) ->
+      let lbl = new_label() in
+      float_literals := (f, lbl) :: !float_literals;
+      `	fld	{emit_reg i.res.(0)}, {emit_label lbl}, {emit_reg reg_tmp1}\n`
+  | Lop(Iconst_symbol s) ->
+      `	la	{emit_reg i.res.(0)}, {emit_symbol s}\n`
+  | Lop(Icall_ind {label_after = label}) ->
+      `	jalr	{emit_reg i.arg.(0)}\n`;
+      record_frame ~label i.live (Dbg_other i.dbg)
+  | Lop(Icall_imm {func; label_after = label}) ->
+      `	{emit_call func}\n`;
+      record_frame ~label i.live (Dbg_other i.dbg)
+  | Lop(Itailcall_ind {label_after = _}) ->
+      let n = frame_size() in
+      if !contains_calls then reload_ra n;
+      emit_stack_adjustment n;
+      `	jr	{emit_reg i.arg.(0)}\n`
+  | Lop(Itailcall_imm {func; label_after = _}) ->
+      if func = !function_name then begin
+        `	j	{emit_label !tailrec_entry_point}\n`
+      end else begin
+        let n = frame_size() in
+        if !contains_calls then reload_ra n;
+        emit_stack_adjustment n;
+        `	{emit_tail func}\n`
+      end
+  | Lop(Iextcall{func; alloc = true; label_after = label}) ->
+      `	la	{emit_reg reg_t2}, {emit_symbol func}\n`;
+      `	{emit_call "caml_c_call"}\n`;
+      record_frame ~label i.live (Dbg_other i.dbg)
+  | Lop(Iextcall{func; alloc = false; label_after = _}) ->
+      `	{emit_call func}\n`
+  | Lop(Istackoffset n) ->
+      assert (n mod 16 = 0);
+      emit_stack_adjustment (-n);
+      stack_offset := !stack_offset + n
+  | Lop(Iload(Single, Iindexed ofs)) ->
+      `	flw	{emit_reg i.res.(0)}, {emit_int ofs}({emit_reg i.arg.(0)})\n`;
+      `	fcvt.d.s	{emit_reg i.res.(0)}, {emit_reg i.res.(0)}\n`
+  | Lop(Iload(chunk, Iindexed ofs)) ->
+      let instr =
+        match chunk with
+        | Byte_unsigned -> "lbu"
+        | Byte_signed -> "lb"
+        | Sixteen_unsigned -> "lhu"
+        | Sixteen_signed -> "lh"
+        | Thirtytwo_unsigned -> if rv64 then "lwu" else "lw"
+        | Thirtytwo_signed -> "lw"
+        | Word_int | Word_val -> if rv64 then "ld" else "lw"
+        | Single -> assert false
+        | Double | Double_u -> "fld"
+      in
+      `	{emit_string instr}	{emit_reg i.res.(0)}, {emit_int ofs}({emit_reg i.arg.(0)})\n`
+  | Lop(Istore(Single, Iindexed ofs, _)) ->
+      `	fmv.x.d	{emit_reg reg_tmp1}, {emit_reg i.arg.(0)}\n`;
+      `	fcvt.s.d	{emit_reg i.arg.(0)}, {emit_reg i.arg.(0)}\n`;
+      `	fsw	{emit_reg i.arg.(0)}, {emit_int ofs}({emit_reg i.arg.(1)})\n`;
+      `	fmv.d.x	{emit_reg i.arg.(0)}, {emit_reg reg_tmp1}\n`
```

Comment:
If we had float-point temporary registers, we could just convert from the source to the temporary and store from there. Don't we have those?

Reply:
Currently all floating-point registers are made available to the register allocator. We could reserve one for "scratch" to use in this case, but is it worth it?

Reply:
This code doesn't work in RV-32 because `fmv.x.d` is not supported.  You can't save a 64-bit FP register in a 32-bit integer register, anyway.  

You don't need to reserve a FP register: you can just use a normal, allocatable register but declare it destroyed by the `Istore` instruction.  You can see this trick at work in the AMD64 port.  Basically, you would emit the following code
```
    `	fcvt.s.d	ft0, {emit_reg i.arg.(0)}\n`;
    `	fsw	ft0, {emit_int ofs}({emit_reg i.arg.(1)})\n`;
```
and add a case to `Proc.destroyed_at_oper`:
```
  | Iop(Istore(Single, _, _)) -> [| phys_reg 100 |]
```

Reply:
After reading more code (I'm not familiar at all with the backend), my impression is that some other backends do this differently: they use a fixed register for the operation (here a floating-point register) and they indicate it in `destroyed_at_oper` for this operation. For example `arm` uses `s14` for single-floating-point load/stores in this way. If I understand correctly, this does not prevent it from being used by the register allocator, just not across this operation.

Reply:
Indeed! I had overlooked this obvious solution.

Reply:
But then, if I understand correctly, you could also use this `destroy_at_oper` for some uses of the "temporary registers" you have today. You use `tmp1` and `tmp2`, the first one is used in many places so it would be an invasive change, but `tmp2` is used only for `Lswitch` so this would be an easy way to free a register.

Reply:
I think you may be right.

For info, `t0` (`reg_tmp1`) is also used in the assembly stubs (`riscv.S`) as a temporary register. Initially also `t1` (`reg_tmp2`) was also used in the assembly stubs, so I didn't mind also using it in the code generator as well, but since the switch to "domain state", the assembly stubs only need one temporary register, so indeed it looks like we may be able to reuse the other one for the register allocator.

I will look into it.

Reply:
Actually after checking the assembly stubs again, it seems `t1` is still used there, so I need to think if indeed the register can be freed for the register allocator.

Reply:
OK, I think it should be possible to reuse `t0` for the register allocator.

This register is not callee-save according to the standard convention, so assumed killed by any C call. In particular, it can be safely overwritten in the routines where it is currently used in the assembly stubs (`caml_callback_exn*_asm`).

I pushed a commit with this change. Thanks!